constantine/ecdsa/ecdsa.nim

```diff
@@ -0,0 +1,356 @@
+  ../math/arithmetic,
+  ../platforms/abstractions,
+  ../serialization/codecs, # for fromHex and (in the future) base64 encoding
+  ../mac/mac_hmac, # for deterministic nonce generation via RFC 6979
+  ../named/zoo_generators, # for generator
+  ../csprngs/sysrand
+
+import std / macros # for `update` convenience helper
+
+type
+  ## Decides the type of sampler we use for the nonce. By default
+  ## a simple uniform random sampler. Alternatively a deterministic
+  ## sampler based on message hash and private key.
+  NonceSampler* = enum
+    nsRandom, ## pure uniform random sampling
+    nsRfc6979 ## deterministic according to RFC 6979
+
+# For easier readibility, define the curve and generator
+# as globals in this file
+const C* = Secp256k1
+const G = Secp256k1.getGenerator("G1")
+
+proc hashMessage(message: string): array[32, byte] =
+  # Hash a given message
+  var h {.noinit.}: sha256
+  h.init()
+  h.update(message)
+  h.finish(result)
+
+proc toBytes(x: Fr[C] | Fp[C]): array[32, byte] =
+  let bi = x.toBig()
+  discard result.marshal(bi, bigEndian)
+
+proc toDER*(r, s: Fr[C]): seq[byte] =
+  ## Converts the given signature `(r, s)` into a signature in
+  ## ASN.1 DER encoding following SEC1.
+  ##
+  ## Note that the implementation is not written for efficiency
+  ## and should be viewed as a convenience tool for the time being.
+  # Convert signature to DER format
+  result = @[byte(0x30)]  # sequence marker
+
+  # Convert r and s to big-endian bytes
+  var rBytes = @(r.toBytes())
+  var sBytes = @(s.toBytes())
+
+  # Add padding if needed (if high bit is set)
+  if (rBytes[0] and 0x80) != 0:
+    rBytes = @[byte(0)] & rBytes
+  if (sBytes[0] and 0x80) != 0:
+    sBytes = @[byte(0)] & sBytes
+
+  # Add integer markers and lengths
+  let rEncoded = @[byte(0x02), byte(rBytes.len)] & rBytes
+  let sEncoded = @[byte(0x02), byte(sBytes.len)] & sBytes
+
+  # Total length
+  let totalLen = rEncoded.len + sEncoded.len
+  result.add(byte(totalLen))
+
+  # Add r and s encodings
+  result.add(rEncoded)
+  result.add(sEncoded)
+
+func fromDigest(dst: var Fr[C], src: array[32, byte]): bool {.discardable.} =
+  ## Convert a SHA256 digest to an element in the scalar field `Fr[Secp256k1]`.
+  ## The proc returns a boolean indicating whether the data in `src` is
+  ## smaller than the field modulus. It is discardable, because in some
+  ## use cases this is fine (e.g. constructing a field element from a hash),
+  ## but invalid in the nonce generation following RFC6979.
+  var scalar {.noInit.}: BigInt[256]
+  scalar.unmarshal(src, bigEndian)
+  # `true` if smaller than modulus
+  result = bool(scalar < Fr[C].getModulus())
+  dst.fromBig(scalar)
+
+proc randomFieldElement[FF](): FF =
+  ## random element in ~Fp[T]/Fr[T]~
+  let m = FF.getModulus()
+  var b: matchingBigInt(FF.Name)
+
+  while b.isZero().bool or (b > m).bool:
+    ## XXX: raise / what else to do if `sysrand` call fails?
+    doAssert b.limbs.sysrand()
```

Comment:
In case this were to fail, do we want a different failure mode?

Reply:
For now it's fine not to address this, AFAIK most protocols use RFC6979 to avoid adding the RNG to the attack surface. If the RNG fails to produce data, I'm not even sure what it means at the OS level, no more entropy maybe?

Reply:
Not sure either to be honest, but given that we _do_ return a `bool` I didn't just want to discard it.